.circleci/config.yml

```
@@ -93,6 +93,75 @@ commands:
             - bytecode-report-<< parameters.label >>-cli.txt
       - matrix_notify_failure_unless_pr
 
+  install_python3:
+    description: "Install python3 and given packages."
+    parameters:
+      packages:
+        description: "List of optional python packages to be installed (separated by space)."
+        type: string
+        default: ""
+    steps:
+      - run:
+          name: Install python and dependencies
+          command: |
+            sudo apt update
+            if ! python3 --version >/dev/null 2>&1
+            then
+              sudo apt install python3 --assume-yes --no-install-recommends
+            fi
+            sudo apt install python3-pip
+            if [ -n "<< parameters.packages >>" ]
+            then
+              echo "Installing additional packages..."
+              python3 -m pip install --user << parameters.packages >>
+            fi
+
+  install_foundry:
+    description: "Install Foundry."
+    parameters:
+      version:
+        type: string
+        default: "nightly"
+      install_path:
+        type: string
+        default: /home/circleci/bin
+    steps:
+      - run:
+          name: Setup Foundry environment variables
+          command: |
+            FOUNDRY_REPO="foundry-rs/foundry"
+            FOUNDRY_VERSION=<< parameters.version >>
+            FOUNDRY_RELEASE_SHA=$(curl \
+              --silent \
+              --fail \
+              --show-error \
+              "https://api.github.com/repos/${FOUNDRY_REPO}/git/refs/tags/${FOUNDRY_VERSION}" \
+              | jq --raw-output .object.sha \
+            )
+            echo "export FOUNDRY_REPO=$FOUNDRY_REPO" >> "$BASH_ENV"
+            echo "export FOUNDRY_VERSION=$FOUNDRY_VERSION" >> "$BASH_ENV"
+            echo "export FOUNDRY_RELEASE_TAG='nightly-${FOUNDRY_RELEASE_SHA}'" >> "$BASH_ENV"
+            # Save commit sha for caching
+            echo $FOUNDRY_RELEASE_SHA > /tmp/workspace/foundry-release-sha
+      - restore_cache:
+          keys:
+            - foundry-<< parameters.version >>-{{ checksum "/tmp/workspace/foundry-release-sha" }}
+      - run:
+          name: Install foundry
+          command: |
+            if ! command -v forge > /dev/null 2>&1; then
```

This step is never supposed to run if the cache gets restored, so I think we can assume foundry is not installed.

Or you could leave just a small sanity check in case we switch images and this turns out to be wrong:
```suggestion
            ! forge --version 2> /dev/null
```